woocommerce/woocommerce
# Prevent orders being placed when no shipping options are available

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

Adds a check to the `OrderController`'s `validate_selected_shipping_methods` method to check if the selected shipping method is a valid method.

In https://github.com/woocommerce/woocommerce/pull/43869/files#diff-cdc9f1829f52d7b83e3542948622d924ef5cf6a3e0b4580e6ebd1c9643bc49b4R501 we made some changes to shipping method handling to prevent fatal errors, the chosen method was cast to a string in this PR.

This meant our check in `validate_selected_shipping_methods` no longer worked because we are checking for: `if ( false === $chosen_shipping_method )` which would no longer be the case.

In this PR I am validating a little further by checking the selected method against the available methods. 

We set the default value of the `chosen_shipping_method` session variable to `[]` when validating orders. This means if it's unset, and shipping is _required_, sessions with no value will not be allowed to sneak through.

This PR brings some changes to PHP unit tests. In the Checkout route unit tests, we now add a Flat Rate shipping method before each test (and remove it afterwards). This is because previously we were using the legacy flat rate which didn't need to be part of a zone to work. This meant that it was not possible to test the fix for this bug as it requires a zone to have a (disabled) shipping method.

We also discovered an issue with the `test_checkout_force_create_account`, `test_checkout_do_not_create_account`, and `test_checkout_create_account` tests. They were passing even though no valid shipping method was selected. This PR fixes that by introducing a new class to override `WC()->session` called. `MockSessionHandler`.

In `MockSessionHandler` I changed places where `wp_set_cache` was used and implemented a class property to act as the cache, and also overrode the `set` function to automatically update the cache when setting data. This was required as the `$this->_data` was overwritten with blank data from the empty cache when logging in as a user. The session was never saved to the DB as the PHPUnit tests run in a single context, so by the time `shutdown` runs the test is over.

In these tests that use it, we are explicitly overriding WC()->session as there is no good way of setting this up using the `woocommerce_session_handler` filter on a per-test basis.

These tests need special handling as they need to be able to set a cookie during guest -> logged in user switching.





Closes https://github.com/woocommerce/woocommerce/issues/45912.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Have a shipping zone with at least one shipping method, but **do not enable it**.  (If there are no shipping zones/methods at all, Woo will not include a shipping section at all on the checkout page.)
2. Make sure that the checkout page is built with the Checkout Block.
3. As a customer, buy a product and go to the checkout page.
4. Eenter an address that does not support shipping. You'll want to see the message "_There are no shipping options available. Please check your shipping address._"
5. Select the **Place Order** button.
6. Ensure the order does not place.

#### Regression testing

1. Enable various shipping zones and various methods (across different countries).
2. Place an order with them all and ensure no errors are encountered.




### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
